### PR TITLE
Pipe tools

### PR DESCRIPTION
**Pull request by jablonskim**

### CODE REVIEW

File: Common/utils.cpp [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+{
+    int buf_size = MultiByteToWideChar(CP_ACP, MB_PRECOMPOSED, str.c_str(), -1, NULL, 0);
+    if (buf_size == 0) {
+        throw std::runtime_error("Error converting string: " + Utils::GetFormattedWindowsErrorMsg());
+    }
+
+    wchar_t* wide_name = new wchar_t[buf_size];
+
+    if (!MultiByteToWideChar(CP_ACP, MB_PRECOMPOSED, str.c_str(), -1, wide_name, buf_size)) {
+         throw std::runtime_error("Error converting string: " + Utils::GetFormattedWindowsErrorMsg());
```
Comment (m-kostrzewa) [inline Common/utils.cpp]: Memory leak here on `wide_name` if exception is thrown.

File: PipeTestTool/pipetesttool.cpp [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+PipeTestTool::PipeTestTool(const wchar_t *pipe_name, unsigned long timeout_ms) :
+    stream(ios),
+    timer(ios),
+    timeout_ms(timeout_ms)
+{
+    DWORD access_flags = GENERIC_READ | GENERIC_WRITE;
+    DWORD attrs = OPEN_EXISTING;
+    DWORD flags = FILE_FLAG_OVERLAPPED;
+
+    HANDLE handle = CreateFile(pipe_name, access_flags, 0, NULL, attrs, flags, NULL);
```
Comment (m-kostrzewa) [inline PipeTestTool/pipetesttool.cpp]: I think we should make all these variables const

File: PipeTestTool/PipeTestTool.qbs [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+        "pipetesttool.cpp",
+        "pipetesttool.h",
+        "../Common/utils.cpp",
+        "../Common/utils.h",
+    ]
+
+    cpp.includePaths: [BOOST_ROOT, "../Common/"]
+    cpp.libraryPaths: [BOOST_ROOT + "/stage/lib/"]
+    cpp.runtimeLibrary: "static"
+    cpp.cxxLanguageVersion: "c++14"
```
Comment (m-kostrzewa) [inline PipeTestTool/PipeTestTool.qbs]: Does it build on 11? If so, I think we should downgrade

File: PipeTestTool/pipetesttool.cpp [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+}
+
+void PipeTestTool::WriteFakeData(size_t echo_len)
+{
+    CopyDataToBuff(FakeData::FakeEtherHdr);
+    CopyDataToBuff(FakeData::FakeAgentHdr);
+
+    size_t headers_len = sizeof(FakeData::ether_header) + sizeof(FakeData::agent_hdr);
+    if (echo_len >= headers_len) {
+        size_t len = echo_len - headers_len;
```
Comment (m-kostrzewa) [inline PipeTestTool/pipetesttool.cpp]: `headers_len` and `len` should be const I think

File: PipeTestTool/pipetesttool.h [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    void AsyncTimer();
+    void ReadHandler(const boost::system::error_code& error, std::size_t bytes_transferred);
+    void WriteHandler(const boost::system::error_code& error, std::size_t bytes_transferred);
+    void TimeoutHandler();
+    void WriteFakeData(size_t echo_len);
+
+    template <typename T>
+    void CopyDataToBuff(const T &data)
+    {
+        auto size = sizeof(data);
```
Comment (m-kostrzewa) [inline PipeTestTool/pipetesttool.h]: Should be const

File: SimplePipeMitmTool/pipemitmtool.cpp [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+}
+
+void PipeMitmTool::PipeContext::WriteHandler(const boost::system::error_code &error, size_t bytes_transferred)
+{
+    if (error) {
+        throw std::runtime_error("[WriteHandler] " + error.message());
+    }
+
+    tx_buff.consume(bytes_transferred);
+    writing = false;
```
Comment (m-kostrzewa) [inline SimplePipeMitmTool/pipemitmtool.cpp:128]: Shouldn't `writing` be std::atomic? Or are we not concerned with data races?

**Review by m-kostrzewa — CHANGES_REQUESTED** [review]
Pretty nice code.

I haven't commented everywhere, where `const` should be used, but consider adding it wherever you can.

File: SimplePipeMitmTool/pipemitmtool.cpp (hunk 6 above)
Comment (jablonskim) [reply SimplePipeMitmTool/pipemitmtool.cpp:128]: It's single-threaded application

Comment (m-kostrzewa) [reply SimplePipeMitmTool/pipemitmtool.cpp:128]: I thought asio handlers could be executed in parallel, but apparently not  https://stackoverflow.com/questions/16066007/does-boost-asio-call-async-handlers-sequentially

**Review by m-kostrzewa — APPROVED** [review]

**Pull request merged**